### PR TITLE
Peer discovery docs

### PR DESCRIPTION
**Pull request by Frederik-Baetens**
Draft, waiting on https://github.com/libp2p/docs/pull/128 so I can refer there for an explanation about boot nodes instead of explaining what boot nodes are in multiple places.

### CODE REVIEW

**Comment by Frederik-Baetens** [comment]
Some of the documentation links don't work when building documentation at the lowest level (libp2p_kad or libp2p_gossipsub) but building them at the root makes all the links work. I don't know how/if to fix this.

**Comment by Frederik-Baetens** [comment]
Rebased & added the link to the boot node explanation. Pr is ready for review now.

**Comment by Frederik-Baetens** [comment]
I would like some advice on how to fix my intra crate documentation links though. They work when the docs are built on the top level, but not when the docs are built for a specific crate like libp2p_kad. (see failed CI run for example)

File: protocols/gossipsub/src/lib.rs [hunk 1]
```diff
@@ -51,6 +51,18 @@
 //! integers. They are chosen at random in this implementation of gossipsub, but are sequential in
 //! the current go implementation.
 //!
+//! # Peer Discovery
+//!
+//! Gossipsub does not provide peer discovery by itself. Peer discovery is the process by which
+//! peers in a p2p network exchange information about eachother in order to become resistant
+//! against the failure or replacement of your
+//! [boot nodes](https://docs.libp2p.io/reference/glossary/#boot-node).
```
Comment (mxinden) [inline protocols/gossipsub/src/lib.rs]:
```suggestion
//! peers in a p2p network exchange information about each other among other reasons to become resistant
//! against the failure or replacement of the
//! [boot nodes](https://docs.libp2p.io/reference/glossary/#boot-node) of the network.
```

File: protocols/gossipsub/src/lib.rs [hunk 2]
```diff
@@ -51,6 +51,18 @@
...
+//!
+//! If your application needs to be resistant against the failure of your boot nodes, peer
+//! discovery can be implemented with the help of the [Kademlia](libp2p_kad::Kademlia) protocol
+//! in combination with the [Identify](libp2p_identify::Identify) protocol. See the
+//! [Kademlia](libp2p_kad::Kademlia) documentation for more information.
```
Comment (mxinden) [inline protocols/gossipsub/src/lib.rs]:
```suggestion
//! Peer
//! discovery can e.g. be implemented with the help of the [Kademlia](libp2p_kad::Kademlia) protocol
//! in combination with the [Identify](libp2p_identify::Identify) protocol. See the
//! [Kademlia](libp2p_kad::Kademlia) documentation for more information.
```

File: protocols/kad/src/lib.rs [hunk 3]
```diff
@@ -19,6 +19,23 @@
 // DEALINGS IN THE SOFTWARE.
 
 //! Implementation of the libp2p-specific Kademlia protocol.
+//!
+//! # Overview
+//! Kademlia provides a distributed hash table in which data can be stored. Kademlia can also be
+//! used to store information about peers in a p2p network.
```
Comment (mxinden) [inline protocols/kad/src/lib.rs]:
```suggestion
//!
//! See [specification](https://github.com/libp2p/specs/blob/master/kad-dht/README.md) for details.
```

I would rather just link to the specification, serving as a single source of truth.

File: protocols/kad/src/lib.rs [hunk 4]
```diff
@@ -19,6 +19,23 @@
...
+//! - **Peer Discovery with Identify** In other libp2p implementations, the
+//! [Identify](libp2p_identify::Identify) protocol might be seen as a core protocol. Rust-libp2p
+//! tries to stay as generic as possible, and does not make this assumption.
+//! This means that the Identify protocol must be manually hooked up to Kademlia through calls
+//! to [`Kademlia::add_address`].
+//! If you choose not to implement the Identify protocol, and do not provide an alternative peer
+//! discovery mechanism, your kademlia network will not be resistant against the failure of your
+//! [boot nodes](https://docs.libp2p.io/reference/glossary/#boot-node). Without the Identify protocol,
+//! existing nodes in the kademlia network cannot obtain the [Multiaddr](multiaddr::Multiaddr)
+//! of nodes querying them, and thus will not be able to reach out to newer nodes.
```
Comment (mxinden) [inline protocols/kad/src/lib.rs]:
```suggestion
//! If you choose not to use the Identify protocol, and do not provide an alternative peer
//! discovery mechanism, a Kademlia node will not discover nodes beyond the network's
//! [boot nodes](https://docs.libp2p.io/reference/glossary/#boot-node). Without the Identify protocol,
//! existing nodes in the kademlia network cannot obtain the listen addresses
//! of nodes querying them, and thus will not be able to add them to their routing table.
```

**Review by mxinden — COMMENTED** [review]
A couple of suggestions. Thanks for the patch.

**Comment by Frederik-Baetens** [comment]
Thanks for the review & suggestions!


Ci still fails for some intra doc links though, don't know how to fix that

File: protocols/kad/src/lib.rs [hunk 5]
```diff
@@ -19,6 +19,21 @@
 // DEALINGS IN THE SOFTWARE.
 
 //! Implementation of the libp2p-specific Kademlia protocol.
+//!
+//! See [specification](https://github.com/libp2p/specs/blob/master/kad-dht/README.md) for details.
+//!
+//! # Important Discrepancies
+//!
+//! - **Peer Discovery with Identify** In other libp2p implementations, the
+//! [Identify](libp2p_identify::Identify) protocol might be seen as a core protocol. Rust-libp2p
```
Comment (mxinden) [inline protocols/kad/src/lib.rs]:
```suggestion
//! [Identify](https://github.com/libp2p/specs/tree/master/identify) protocol might be seen as a core protocol. Rust-libp2p
```

File: protocols/gossipsub/src/lib.rs [hunk 6]
```diff
@@ -51,6 +51,18 @@
...
+//! # Peer Discovery
+//!
+//! Gossipsub does not provide peer discovery by itself. Peer discovery is the process by which
+//! peers in a p2p network exchange information about each other among other reasons to become resistant
+//! against the failure or replacement of the
+//! [boot nodes](https://docs.libp2p.io/reference/glossary/#boot-node) of the network.
+//!
+//! Peer
+//! discovery can e.g. be implemented with the help of the [Kademlia](libp2p_kad::Kademlia) protocol
+//! in combination with the [Identify](libp2p_identify::Identify) protocol. See the
```
Comment (mxinden) [inline protocols/gossipsub/src/lib.rs]:
```suggestion
//! in combination with the [Identify](https://github.com/libp2p/specs/tree/master/identify) protocol. See the
```

File: protocols/gossipsub/src/lib.rs [hunk 7]
```diff
@@ -51,6 +51,18 @@
...
 //!
+//! # Peer Discovery
+//!
+//! Gossipsub does not provide peer discovery by itself. Peer discovery is the process by which
+//! peers in a p2p network exchange information about each other among other reasons to become resistant
+//! against the failure or replacement of the
+//! [boot nodes](https://docs.libp2p.io/reference/glossary/#boot-node) of the network.
+//!
+//! Peer
+//! discovery can e.g. be implemented with the help of the [Kademlia](libp2p_kad::Kademlia) protocol
```
Comment (mxinden) [inline protocols/gossipsub/src/lib.rs]:
```suggestion
//! discovery can e.g. be implemented with the help of the [Kademlia](https://github.com/libp2p/specs/blob/master/kad-dht/README.md) protocol
```

File: protocols/gossipsub/src/lib.rs [hunk 8]
```diff
@@ -51,6 +51,18 @@
...
+//! in combination with the [Identify](libp2p_identify::Identify) protocol. See the
+//! [Kademlia](libp2p_kad::Kademlia) documentation for more information.
```
Comment (mxinden) [inline protocols/gossipsub/src/lib.rs]:
```suggestion
//! Kademlia implementation documentation for more information.
```

**Review by mxinden — COMMENTED** [review]
Unfortunately one can not link to crates that are not listed in one's `Cargo.toml`.

**Review by mxinden — APPROVED** [review]
Thanks!

**Pull request merged**